index.bs

```diff
@@ -1,17 +1,222 @@
+
+  1. Return |bc|.
+</div>
+
+<div algorithm>
+  To <dfn for="prerendering browsing context">activate</dfn> a [=prerendering browsing context=] |successorBC| in place of a [=top-level browsing context=] |predecessorBC| given a [=history handling behavior=] |historyHandling|:
+
+  1. Assert: |historyHandling| is either "`default`" or "`replace`".
+
+  1. Assert: |successorBC| is not [=prerendering browsing context/empty=].
+
+  1. Assert: |predecessorBC| is a [=top-level browsing context=].
+
+  <!-- The following are copied from the navigate algorithm, and probably could benefit from some refactoring to deduplicate. -->
+
+  1. Cancel any preexisting but not yet [=navigate/mature=] attempts to navigate |predecessorBC|, including canceling any instances of the [=fetch=] algorithm started by those attempts. If one of those attempts has already <a lt="create and initialize a Document object">created and initialized a new `Document` object</a>, [=Document/abort=] that {{Document}} also.
+
+  1. [=Prompt to unload=] the [=active document=] of |predecessorBC|. If the user [=refused to allow the document to be unloaded=], then return.
+
+  1. [=Document/Abort=] the [=active document=] of |predecessorBC|.
+
+  <!-- End copied section. -->
+
+  1. TODO prepend the existing session history of |predecessorBC| into |successorBC|? Or, probably better, use the new "browsing session" concept to bridge them? Be sure to respect |historyHandling|.
+
+  1. [=In parallel=]:
+
+    1. Update the user agent's user interface to replace |predecessorBC| with |successorBC|, e.g., by updating the tab/window contents and the browser chrome.
+
+    <!-- TODO is this the right task source? Should we make a new one? -->
+    1. [=Queue a global task=] on the [=networking task source=] given |successorBC|'s [=browsing context/active window=] to perform the following steps
+
+      1. Set |successorBC|'s [=browsing context/loading mode=] to "`default`".
+</div>
+
+<h2 id="navigation">Navigation and session history</h2>
+
+<h3 id="navigate-activation">Allowing activation in place of navigation</h3>
+
+Patch the [=navigate=] algorithm to allow the [=prerendering browsing context/activate|activation=] of a [=prerendering browsing context=] in place of a normal navigation as follows:
+
+<div algorithm="navigate activate patch">
+  In [=navigate=], append the following steps after the fragment navigation handling (currently step 6):
+
+  1. If all of the following are true:
+
+      * |browsingContext| is a [=top-level browsing context=]
+      * |historyHandling| is "`default`" or "`replace`"
+      * <var ignore>navigationType</var> is "`other`"
+      * |resource| is a [=request=] whose [=request/method=] is \``GET`\`
+      * |browsingContext|'s [=active document=]'s [=Document/prerendering browsing contexts map=][|resource|'s [=request/URL=]] [=map/exists=]
```

Comment:
Are there any other request fields that need to form part of this key? e.g. referrer policy (I think not)?

Reply:
I at first thought not, but now that you bring up referrer policy, I think that does need to be a key. In particular, consider

```
<link rel="prerender2" referrerpolicy="unsafe-url" href="https://example.com">

<a referrerpolicy="no-referrer" href="https://example.com/">
```

I don't think it's a good idea to use the prerendered version from the `<link>` to fulfill any navigations from that `<a>`...